src/firefly/java/edu/caltech/ipac/firefly/server/network/HttpServices.java

```diff
@@ -210,50 +198,88 @@ public static int executeMethod(HttpMethod method, HttpServiceInput input, Outpu
             handleParams(method, input.getParams(), input.getFiles());
 
             int status = httpClient.executeMethod(method);
-
-            handleResults(method, results);
-
-            boolean isSuccess =  status >= 200 && status < 300;
-            if (!isSuccess) {
+            if (status < 200 || status >= 300) {
+                // logs bad requests
                 LOG.error("HTTP request failed with status:" + status + "\n" + getDetailDesc(method, input));
             }
-            return status;
+
+            if (handler != null) {
+                handler.handleResponse(method);
+            }
+
+            return method;
+
         } finally {
-            if (results != null) {
+            if (handler != null) {
                 method.releaseConnection();
             }
         }
     }
 
 
-
-
-
 //====================================================================
 //  Util helper functions
 //====================================================================
 
-    public static void handleResults(HttpMethod method, OutputStream results) {
-        BufferedInputStream bis = null;
-        BufferedOutputStream bos = null;
-        try {
-            String encoding = getResHeader(method, "Content-Encoding");
-            if (encoding.contains("gzip")) {
-                bis = new BufferedInputStream(new GZIPInputStream(method.getResponseBodyAsStream()));
-            } else {
-                bis = new BufferedInputStream(method.getResponseBodyAsStream());
-            }
+    public static Handler defaultHandler(File source) throws FileNotFoundException {
+        return source == null ? null : new OutputStreamHandler(source);
+    }
+
+    public static Handler defaultHandler(OutputStream source) {
+        return source == null ? null : new OutputStreamHandler(source);
+    }
+
+    public static class Status {
+        private String errMsg;
+        private int statusCode;
+
+        public Status(int statusCode, String errMsg) {
+            this.errMsg = errMsg;
+            this.statusCode = statusCode;
+        }
+
+        public boolean isError() { return statusCode < 200 || statusCode >= 300; }
```

Comment:
I like that we now pass Status. I think it's OK to treat 3xx codes (redirects) as an error. If HttpServices is not set to follow redirect, it's up to the user to deal with it.